### PR TITLE
implement Tuple match pattern

### PR DESCRIPTION
**Pull request by StachuDotNet**
This implements a new `match` pattern, to match against tuples, like:

image

Much of both the backend and client work is complete, but there's still a fair amount of client work and testing left.
(esp. misc. off-by-one fluid issues that are bound to come up)

TODOs tracked in #4427.

### CODE REVIEW

**Comment by StachuDotNet** [comment]
for the tuple match pattern, someone could try to bind the same var name in the sub-patterns like:
```
let x = (1, 2, 3)
match x
  | (a, 2, a) -> ...
```

- [ ] think about this. should this be prevented? This doesn't feel like something the interpreter should have to deal with, but rather prevented earlier? And if such a case _does_ end up being Interpreted, how should it handle such? (last pattern wins? what about for nested patterns? hmm)

**Comment by pbiggar** [comment]
> for the tuple match pattern, someone could try to bind the same var name in the sub-patterns like:
> 
> ```
> let x = (1, 2, 3)
> match x
>   | (a, 2, a) -> ...
> ```
> 
> * [ ]  think about this. should this be prevented? This doesn't feel like something the interpreter should have to deal with, but rather prevented earlier? And if such a case _does_ end up being Interpreted, how should it handle such? (last pattern wins? what about for nested patterns? hmm)

I think last pattern should win. I agree we shouldn't raise an error in the interpreter, that's better done in the editor.

**Comment by StachuDotNet** [comment]
When a user's cursor is on an existing Pattern, should a `(` key stroke generally _wrap_ that existing pattern?
I think so.

File: client/src/fluid/FluidTypes.res [hunk 1]
```diff
@@ -127,18 +127,30 @@ module Token = {
...
+
+    | TPatternTupleOpen(ID.t)
+    | TPatternTupleComma(ID.t, int)
+    | TPatternTupleClose(ID.t)
+
     | TConstructorName(ID.t, string)
+
+    // do we even use these?
     | TParenOpen(ID.t)
     | TParenClose(ID.t)
```
Comment (StachuDotNet) [inline client/src/fluid/FluidTypes.res:152]: I noticed TParenOpen and TParenClose - I was wondering when we used these, if ever

File: client/src/fluid/FluidTypes.res [hunk 2]
```diff
@@ -127,18 +127,30 @@ module Token = {
...
+
     | TPatternInteger(ID.t, ID.t, int64, int)
     | TPatternString({matchID: ID.t, patternID: ID.t, str: string, branchIdx: int})
+
     | TPatternTrue(ID.t, ID.t, int)
     | TPatternFalse(ID.t, ID.t, int)
-    | TPatternNullToken(ID.t, ID.t, int)
+
+
+    | TPatternNullToken(ID.t, ID.t, int)// why is 'token' in this name
```
Comment (StachuDotNet) [inline client/src/fluid/FluidTypes.res:138]: If needless, I'll remove the "token" bit in a tiny PR

File: client/src/fluid/FluidTokenizer.res [hunk 3]
```diff
@@ -166,6 +166,27 @@ let rec patternToToken = (matchID: id, p: FluidPattern.t, ~idx: int): list<fluid
...
+
+        if isLastPattern {
+          subpatternTokens
+        } else {
+          List.append(subpatternTokens, list{TPatternTupleComma(id, i)})
+        }
+      })
+      |> List.flatten
+
+    List.flatten(list{list{TPatternTupleOpen(id)}, middlePart, list{TPatternTupleClose(id)}})
```
Comment (StachuDotNet) [inline client/src/fluid/FluidTokenizer.res:189]: Am I missing some really obvious way of going from:
`[patternA, patternB, patternC]` to `[patternATokens, tupleCommaToken, patternBTokens, tupleCommaToken, patternCTokens]`?

Comment (StachuDotNet) [reply client/src/fluid/FluidTokenizer.res:189]: it feels like there should be a more terse option, but I'm blanking.

File: client/src/fluid/Fluid.res [hunk 4]
```diff
@@ -33,7 +33,7 @@ open ProgramTypes.Expr
 open ProgramTypes.Pattern
 
 // TUPLETODO re-enable
-let allowUserToCreateTuple = false
+let allowUserToCreateTuple = true
```
Comment (StachuDotNet) [inline client/src/fluid/Fluid.res:36]: TODO: flip this back.

**Comment by StachuDotNet** [comment]
I think the TODOs noted in the OP of this PR will take at least another day or two until complete.

That said, this gets the general idea out
- the tuple match pattern _type_ is defined
- the interpreter respects the new pattern, and basic tests pass
- Fluid has been updated to add new new tuple pattern to a `match` (though only tuples of size 2, currently), and have its sub-patterns filled

So, I think this _could_ be nearly merged, splitting off other work as follow-up TODOs.
(after 'hiding' the creation of tuple patterns behind the `allowUserToCreateTuple` flag)

What do you think - is that worth pursuing, or should this PR only be merged once everything is solid?

**Comment by StachuDotNet** [comment]
Here's a short demo of progress:
https://www.youtube.com/watch?v=HoK1sxYYbcE

File: client/src/fluid/FluidTypes.res (hunk 1 above)
Comment (pbiggar) [reply client/src/fluid/FluidTypes.res:152]: There is a nearly complete implementation of precedence in the editor, this is part of it.

**Comment by pbiggar** [comment]
> When a user's cursor is on an existing Pattern, should a `(` key stroke generally _wrap_ that existing pattern? I think so.

I think it should mimic the expr behavior

File: client/src/fluid/FluidTypes.res (hunk 2 above)
Comment (pbiggar) [reply client/src/fluid/FluidTypes.res:138]: Yeah, seems useless

File: client/src/fluid/FluidTokenizer.res (hunk 3 above)
Comment (pbiggar) [reply client/src/fluid/FluidTokenizer.res:189]: I thought that's what `intersperse` was for

**Comment by pbiggar** [comment]
> What do you think - is that worth pursuing, or should this PR only be merged once everything is solid?

Given it's behind a flag, we have quite some flexibility with this. Better to merge incomplete work behind a flag, rather than having huge PRs, imo.

File: client/src/fluid/FluidTokenizer.res (hunk 3 above)
Comment (StachuDotNet) [reply client/src/fluid/FluidTokenizer.res:189]: The `TPatternTupleComma`s include an index `i`. I'm not sure how to include the context of the `Comma`'s index using `List.intersperse`

Maybe I want a `List.intersperceWithIndex` :thinking:?

**Comment by StachuDotNet** [comment]
> > What do you think - is that worth pursuing, or should this PR only be merged once everything is solid?
> 
> Given it's behind a flag, we have quite some flexibility with this. Better to merge incomplete work behind a flag, rather than having huge PRs, imo.

Agreed. I think the PR's in a good state to be reviewed/merged now, in that case.

File: client/src/fluid/FluidShortcuts.res [hunk 5]
```diff
@@ -138,6 +138,14 @@ let pNull = (~id=gid(), ()): FluidPattern.t => PNull(id)
 
 let pBlank = (~id=gid(), ()): FluidPattern.t => PBlank(id)
 
+// TUPLETODO use this in tests
```
Comment (pbiggar) [inline client/src/fluid/FluidShortcuts.res:141]: Outstanding todo, FYI

File: client/src/fluid/FluidTokenizer.res [hunk 6]
```diff
@@ -166,6 +167,33 @@ let rec patternToToken = (matchID: id, p: FluidPattern.t, ~idx: int): list<fluid
     Belt.List.concatMany([whole, list{TPatternFloatPoint(matchID, id, idx)}, fraction])
   | PNull(id) => list{TPatternNullToken(matchID, id, idx)}
   | PBlank(id) => list{TPatternBlank(matchID, id, idx)}
+  | PTuple(id, first, second, theRest) =>
+    let subPatterns = list{first, second, ...theRest}
+
+    // CLEANUP: this is likely not the most efficient way to handle this.
```
Comment (pbiggar) [inline client/src/fluid/FluidTokenizer.res:173]: Performance wise, this seems totally fine. As a result I would remove the comment as I think it'll be a red herring. They only performance improvements I would recommend here is to pull out of List.length as it is an invariant.

File: fsharp-backend/src/LibExecution/Interpreter.fs [hunk 7]
```diff
@@ -278,16 +282,39 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
...
+          | DTuple (first, second, theRest) ->
+            let allVals = first :: second :: theRest
+
+            if List.length allVals = List.length allPatterns then
+              let (passResults, newVarResults, traceResults) =
+                List.zip allVals allPatterns
+                |> List.map (fun (dv, pat) -> checkPattern dv pat)
+                |> List.unzip3
+
+              let allPass = passResults |> List.forall (fun arg -> arg)
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/Interpreter.fs:306]: There should be a function called `identity` somewhere that we can use for this.

**Review by pbiggar — APPROVED** [review]
Very nice!

File: client/src/fluid/FluidShortcuts.res (hunk 5 above)
Comment (StachuDotNet) [reply client/src/fluid/FluidShortcuts.res:141]: Thanks - I'll do this in future PRs, searching throughout for TUPLETODO before marking tuples as no-longer in-progress.

File: client/src/fluid/FluidTokenizer.res (hunk 6 above)
Comment (StachuDotNet) [reply client/src/fluid/FluidTokenizer.res]: Yeah I agree the perf is fine here - I was more looking for terser code, but again this is fine. :heavy_check_mark: w/r/t List.length

File: fsharp-backend/src/LibExecution/Interpreter.fs (hunk 7 above)
Comment (StachuDotNet) [reply fsharp-backend/src/LibExecution/Interpreter.fs:306]: Found it - `Fun.identity`!

**Comment by StachuDotNet** [comment]
Merging this as-is; if I did the wrong thing putting `identity` into Prelude, we can tidy that in a small follow-up PR.

**Pull request merged**